### ISSUE TITLE
Provide info about underlying connection in Request

### ISSUE THREAD

**Issue by little-arhat** [mirage/ocaml-cohttp#188]
In some cases one may want to access info about underlying connection in request handler function, such as ip address of peer, local port. 

There is no obvious way to do so currently.

It would be nice to have such features in cohttp.

**Comment by gasche** [comment]
Two more people have asked about this feature on the #ocaml IRC channel yesterday, Algebr and @AltGr .

**Comment by yomimono** [comment]
One more person has asked about this feature on the issue tracker today, and it's me. I want to implement some rate-limiting per remote IP for a webserver, and if my type archaeology is correct, that's still not possible. :cry:

**Comment by mseri** [comment]
Yes. This is still not possible. If I recall correctly this would need some changes at conduit level. We are breaking some apis, any contribution is welcome

**Comment by paurkedal** [comment]
Make that +1. My use cases is that I made a [unikernel](https://github.com/paurkedal/ephca) which issues unvalidated certificates, merely identified by serial number. It primarily meant for CIs or other throw-away environments, though it would be nice if it could validate source IP or domain name.

**Comment by yomimono** [comment]
It seems like some relevant discussion is in https://github.com/mirage/ocaml-cohttp/issues/779 ("cohttp API 4 questions") but it's rather divorced from what I think is also an ongoing conversation in https://github.com/mirage/ocaml-cohttp/issues/821 ("Conduit Free Servers") , and I know there's additional context around MirageOS 4.0 and the future of web server invocation there.

I don't have my head enough in the game to make a concrete suggestion -- I already missed my chance to fix this once! But big refactors are a good time to make sure your wheel squeaks enough not to be forgotten, even if it is more annoying noise for maintainers who hear nothing but wheel squeaks all day :/

**Comment by mseri** [comment]
@rgrinberg do you have an idea on how to address this in a clean way?

**Comment by hannesm** [comment]
I'm neither @rgrinberg nor do I know what a clean way would be, but one possibility is https://github.com/roburio/ocaml-conduit/tree/client-ip4 and https://github.com/roburio/ocaml-cohttp/tree/client-ip4 (that's on top of the 4.0 release). This breaks the API, but has been shown useful for my use cases.

**Comment by dinosaure** [comment]
On the other side, I'm not sure that the usage of `conduit` is relevant for the server part. For instance, [`paf`](https://github.com/dinosaure/paf-le-chien) provides the underlying flow used to communicate with clients.
